File: src/Core.CSharp.Observe/Algebra.cs
namespace Zeta.Core.CSharp.Observe;

/// <summary>
/// The pure event algebra. Three functions, all deterministic + side-effect-free:
/// <c>Simulate</c> (the reducer / step), <c>Fold</c> (left-fold = projection),
/// <c>Replay</c> (state after each event). These mirror tools/observe/observe.ts
/// and src/Core.FSharp.Observe byte-for-byte in behaviour, so the shared
/// golden-vector fixture produces the SAME states here — the cross-language-parity
/// = non-Byzantine-BFT check (B-0944).
///
/// Active disciplines: lock-free + wait-free (pure values, no shared mutable
/// state), weight-free (no implicit weighting), DST (deterministic from the same
/// event log), idempotency (fold = the monoidal reduction over an append-only log).
/// </summary>
public static class Algebra
{
    /// <summary>Apply one action to the world, returning the next world. The
    /// <c>Reason</c> fields never affect the transition (parity with TS/F#).</summary>
    public static World Simulate(World world, NextAction action) => action switch
    {
        // Clear the pending-ferry signal (only if the channel is wired).
        NextAction.PreserveFerry =>
            world.Operator is { } op1
                ? world with { Operator = op1 with { PendingFerry = false } }
                : world,

        // Clear the pending-message signal (only if the channel is wired).
        NextAction.RespondToOperator =>
            world.Operator is { } op2
                ? world with { Operator = op2 with { PendingMessage = false } }
                : world,

        // Item done → drop it from the backlog; entering work mode.
        NextAction.DoItem(var item) =>
            world with
            {
                Backlog = world.Backlog.Where(i => !string.Equals(i.Id, item.Id, StringComparison.Ordinal)).ToList(),
                Mode = Mode.Work,
            },

        // Replace the ambiguous item in place with two ready children.
        NextAction.Decompose(var item) =>
            world with
            {
                Backlog = world.Backlog
                    .SelectMany(i => string.Equals(i.Id, item.Id, StringComparison.Ordinal)
                        ? new[]
                        {
                            new BacklogItem(item.Id + ".1", item.Title + " (part 1)", Ready: true, Ambiguous: false, NeedsNewAction: false),
                            new BacklogItem(item.Id + ".2", item.Title + " (part 2)", Ready: true, Ambiguous: false, NeedsNewAction: false),
                        }
                        : new[] { i })
                    .ToList(),
                Mode = Mode.Work,
            },

        // Sovereign grammar edit: the targeted item gains a fresh action + becomes
        // ready (no-op if no target item). Entering work mode.
        NextAction.EditGrammar(var item, _) =>
            item is null
                ? world
                : world with
                {
                    Backlog = world.Backlog
                        .Select(i => string.Equals(i.Id, item.Id, StringComparison.Ordinal)
                            ? i with { NeedsNewAction = false, Ready = true, Ambiguous = false }
                            : i)
                        .ToList(),
                    Mode = Mode.Work,
                },

        // The four free modes set the persisted mode and leave the backlog alone.
        NextAction.Explore => world with { Mode = Mode.Explore },
        NextAction.Play => world with { Mode = Mode.Play },
        NextAction.SelfReflect => world with { Mode = Mode.SelfReflect },
        NextAction.FreeTime => world with { Mode = Mode.FreeTime },

        _ => throw new ArgumentOutOfRangeException(nameof(action), action, "unknown NextAction"),
    };

    /// <summary>Project state from the event log: left-fold over <c>Simulate</c>.
    /// History is a list of events; state is a projection of that list.</summary>
    public static World Fold(World initial, IEnumerable<NextAction> events) =>
        events.Aggregate(initial, Simulate);

    /// <summary>State after each event, the initial state excluded — mirrors the
    /// TS/F# <c>replay</c>.</summary>
    public static IReadOnlyList<World> Replay(World initial, IEnumerable<NextAction> events)
    {
        var states = new List<World>();
        var current = initial;
        foreach (var ev in events)
        {
            current = Simulate(current, ev);
            states.Add(current);
        }

        return states;
    }
}

File: src/Core.Rust.Observe/src/algebra.rs
//! The pure event algebra: `simulate` (reducer) / `fold` (projection) / `replay`.
//! Mirrors tools/observe/observe.ts + the F#/C# oracles byte-for-byte in behaviour,
//! so the shared golden-vector fixture produces the SAME states here — the
//! cross-language-parity = non-Byzantine-BFT check (B-0944).
//!
//! Active disciplines: lock-free + wait-free (pure values, no shared mutable state),
//! weight-free, DST (deterministic from the same event log), idempotency (fold =
//! the monoidal reduction over an append-only log).

use crate::types::{BacklogItem, Mode, NextAction, OperatorChannel, World};

/// Apply one action to the world, returning the next world. `reason` fields never
/// affect the transition (parity with TS/F#/C#).
pub fn simulate(world: &World, action: &NextAction) -> World {
    match action {
        // Clear the pending-ferry signal (only if the channel is wired).
        NextAction::PreserveFerry { .. } => World {
            backlog: world.backlog.clone(),
            operator: world.operator.as_ref().map(|op| OperatorChannel {
                pending_message: op.pending_message,
                pending_ferry: false,
            }),
            mode: world.mode,
        },

        // Clear the pending-message signal (only if the channel is wired).
        NextAction::RespondToOperator { .. } => World {
            backlog: world.backlog.clone(),
            operator: world.operator.as_ref().map(|op| OperatorChannel {
                pending_message: false,
                pending_ferry: op.pending_ferry,
            }),
            mode: world.mode,
        },

        // Item done → drop it from the backlog; entering work mode.
        NextAction::DoItem { item } => World {
            backlog: world.backlog.iter().filter(|i| i.id != item.id).cloned().collect(),
            operator: world.operator.clone(),
            mode: Some(Mode::Work),
        },

        // Replace the ambiguous item in place with two ready children.
        NextAction::Decompose { item } => World {
            backlog: world
                .backlog
                .iter()
                .flat_map(|i| {
                    if i.id == item.id {
                        vec![
                            BacklogItem {
                                id: format!("{}.1", item.id),
                                title: format!("{} (part 1)", item.title),
                                ready: true,
                                ambiguous: false,
                                needs_new_action: false,
                            },
                            BacklogItem {
                                id: format!("{}.2", item.id),
                                title: format!("{} (part 2)", item.title),
                                ready: true,
                                ambiguous: false,
                                needs_new_action: false,
                            },
                        ]
                    } else {
                        vec![i.clone()]
                    }
                })
                .collect(),
            operator: world.operator.clone(),
            mode: Some(Mode::Work),
        },

        // Sovereign grammar edit: the targeted item gains a fresh action + becomes
        // ready (no-op if no target item). Entering work mode.
        NextAction::EditGrammar { item, .. } => match item {
            None => world.clone(),
            Some(target) => World {
                backlog: world
                    .backlog
                    .iter()
                    .map(|i| {
                        if i.id == target.id {
                            BacklogItem {
                                ready: true,
                                ambiguous: false,
                                needs_new_action: false,
                                ..i.clone()
                            }
                        } else {
                            i.clone()
                        }
                    })
                    .collect(),
                operator: world.operator.clone(),
                mode: Some(Mode::Work),
            },
        },

        // The four free modes set the persisted mode and leave the backlog alone.
        NextAction::Explore { .. } => World { mode: Some(Mode::Explore), ..world.clone() },
        NextAction::Play { .. } => World { mode: Some(Mode::Play), ..world.clone() },
        NextAction::SelfReflect { .. } => World { mode: Some(Mode::SelfReflect), ..world.clone() },
        NextAction::FreeTime { .. } => World { mode: Some(Mode::FreeTime), ..world.clone() },
    }
}

/// Project state from the event log: left-fold over `simulate`.
/// History is a list of events; state is a projection of that list.
pub fn fold(initial: &World, events: &[NextAction]) -> World {
    events.iter().fold(initial.clone(), |w, a| simulate(&w, a))
}

/// State after each event, the initial state excluded — mirrors the TS/F#/C# `replay`.
pub fn replay(initial: &World, events: &[NextAction]) -> Vec<World> {
    let mut states = Vec::with_capacity(events.len());
    let mut current = initial.clone();
    for a in events {
        current = simulate(&current, a);
        states.push(current.clone());
    }
    states
}
